src/hazelcore/HTyp.rei

```diff
@@ -1,7 +1,17 @@
 /* types with holes */
+
+[@deriving sexp]
+type hole_provenance =
+  | SynPatternVar
+  /* comes from a pattern var being asked to synthesize a type either directly or via matched arrow/prod/etc.
+     When analyzing against such a type, expression can be treated as if synthetic when the distinction matters
+     e.g. for inconsistent branches or cast insertion */
+  | Internal /* other internally generated unknown types, e.g. the type synthesized by a hole in synthetic position etc. */
+  | TypHole; /* comes from an actual type hole, will add an `of MetaVar.t`, once we have unique IDs for type holes */
```

Comment:
for readability lets make this the first constructor of the three and Internal the last